pallets/stateful-storage/src/stateful_signed_extension.rs

```diff
@@ -0,0 +1,281 @@
+	DispatchError,
+};
+
+/// The SignedExtension trait is implemented on CheckFreeExtrinsicUse to validate that a provider
+/// has not already been revoked if the calling extrinsic is revoking a provider to an MSA. The
+/// purpose of this is to ensure that the revoke_delegation_by_delegator extrinsic cannot be
+/// repeatedly called and flood the network.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct StatefulSignedExtension<T: Config + Send + Sync>(PhantomData<T>);
+
+pub trait StatefulStorageValidate<C: Config> {
+	fn validate_itemized(
+		msa_id: &MessageSourceId,
+		schema_id: &SchemaId,
+		actions: &BoundedVec<ItemAction, C::MaxItemizedActionsCount>,
+		is_payload_signed: bool,
+		target_hash: &PageHash,
+	) -> TransactionValidity;
+
+	fn validate_paginated(
+		msa_id: &MessageSourceId,
+		schema_id: &SchemaId,
+		page_id: &PageId,
+		is_payload_signed: bool,
+		page_delete: bool,
+		target_hash: &PageHash,
+	) -> TransactionValidity;
+}
+
+impl<T: Config> StatefulStorageValidate<T> for Pallet<T> {
+	/// Validates the following criteria for an Itemized Stateful Storage extrinsic:
+	///
+	/// * schema_id represents a valid schema
+	/// * schema `PayloadLocation` is `Paginated`
+	/// * schema allows signed payloads if `is_payload_signed` == true
+	/// * schema is not `AppendOnly` if `page_delete` == true
+	/// * page can be read/parsed
+	/// * known page state hash matches provided target hash
+	///
+	/// Returns a `ValidTransaction` or wrapped [`pallet::Error`]
+	///
+	/// # Arguments:
+	/// * account_id: the account id associated with the MSA owning the storage
+	/// * msa_id: the `MessageSourceAccount`
+	/// * schema_id: the `SchemaId` for the Paginated storage
+	/// * actions: Array of `ItemAction` representing adds/deletes to Itemized storage
+	/// * is_payload_signed: whether we were called with a signed payload
+	/// * target_hash: the content hash representing the known state of the Paginated storage.
+	///
+	/// # Errors (as u8 wrapped by `InvalidTransaction::Custom`)
+	/// * [`Error::InvalidSchemaId`]
+	/// * [`Error::SchemaPayloadLocationMismatch`]
+	/// * [`Error::SchemaNotSupported`]
+	/// * [`Error::CorruptedState`]
+	/// * [`Error::StalePageState`]
+	///
+	fn validate_itemized(
+		msa_id: &MessageSourceId,
+		schema_id: &SchemaId,
+		actions: &BoundedVec<ItemAction, T::MaxItemizedActionsCount>,
+		is_payload_signed: bool,
+		target_hash: &PageHash,
+	) -> TransactionValidity {
+		const TAG_PREFIX: &str = "StatefulStorageItemized";
+		let is_pruning = actions.iter().any(|a| matches!(a, ItemAction::Delete { .. }));
+
+		Self::check_schema(*schema_id, PayloadLocation::Itemized, is_payload_signed, is_pruning)
+			.map_err(|e| map_dispatch_error(e))?;
+
+		Self::get_checked_itemized_page(msa_id, schema_id, target_hash)
+			.map_err(|e| map_dispatch_error(e))?;
+
+		return ValidTransaction::with_tag_prefix(TAG_PREFIX)
+			.and_provides(msa_id)
```

Comment:
I think all of these should be combined (maybe in a tuple) and only use one `provides` call. Looking at `and_provides` implementation is seems like it treats each as a prefix.

Reply:
Looking at `ValidTransaction` documents it seems like we should include msa_id and schema_id in our prefix for itemized and msa_id, schema_id and page_id in our paginated and get rid of provides.
Provides and requires are there to allow creating a dependency graph between transactions.

Reply:
Updated.

Reply:
Looking at the code, the prefix is not part of the `ValidTransaction`; it only gets used in the builder when you add a tag with either `provides` or `requires`.